### PR TITLE
DS-3428 - Add closed groups

### PR DESCRIPTION
**Pull request by jaapjan**
## HTT - Setup

- [x] Check out the [code](https://github.com/goalgorilla/open_social/pull/381/files)
- [x] Check out this branch `git checkout feature/DS-3428-closed-groups`
- [x] Do a `drush fra -y --bundle=social; drush updb -y; drush entup -y; drush cr`

## HTT - Create closed group and content

- [x] Login as LU and click the [+], select `"New group"`
- [x] Notice you can choose between `"Open group"` and `"Closed group"` (Choose closed group)
- [x] Fill out the group details and save the closed group
- [x] Post something to the group stream
- [x] Create a new topic and event in this closed group
- [x] Check out `"/all-topics"`, `"/all-events"` and the search, since you are a member of the group notice you are able to view all the content from the closed group platform wide.

## HTT - Non-member / outsider

- [x] Login as another LU that is not a member of the closed group
- [x] On the homepage `"/stream"` notice you don't see any activity items in the stream from the closed group
- [x] Go to `"/all-groups"` and notice the closed group
- [x] Click the closed group
- [x] Notice how you end up at the about page of the group and that you can not see any events, topics or posts (stream)
- [x] Notice you can not join the group yourself
- [x] Notice you can see the group manager and members page
- [x] In the hero area, notice that it is a closed group
- [x] Go to `"/all-topics"` and `"/all-events"` and notice you can't see any of the topics/events you've created inside the closed group
- [x] Fill out the search with the title of closed group content and notice you don't get any results

## HTT - Delete closed group and all of its content

- [x] Login as the LU that created the closed group
- [x] Go to "/group/x/edit"
- [x] Notice that upon edit it is not allowed to change it to an open group (or vice versa)
- [x] Delete this group
- [x] Notice you see that the group was deleted with a message and you don't see any of your group content across the platform

### CODE REVIEW

File: modules/social_features/social_mentions/src/Plugin/ActivityContext/MentionActivityContext.php [hunk 1]
```diff
@@ -45,10 +47,21 @@ public function getRecipients(array $data, $last_uid, $limit) {
...
-              'target_id' => $mention->getMentionedUserId(),
-            ];
+            $uid = $mention->getMentionedUserId();
+
+            $entity_storage = \Drupal::entityTypeManager()
+              ->getStorage($mention->getMentionedEntityTypeId());
+            $mentioned_entity = $entity_storage->load($mention->getMentionedEntityId());
+
+            $account = User::load($uid);
+            $access = $mentioned_entity->access('view', $account);
```
Comment (jaapjan) [inline modules/social_features/social_mentions/src/Plugin/ActivityContext/MentionActivityContext.php]: @jochemvn especially check out this code! Like i mentioned in the DS this morning we should also check for view access on the entity in which the user is mentioned before adding them as a recipient.

File: tests/behat/features/capabilities/group/group-create-closed.feature [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+  Role: As a LU
+  Goal/desire: I want to create Closed Groups
+
+  Scenario: Successfully create closed group
+    Given users:
+      | name           | mail                     | status |
+      | Group User One | group_user_1@example.com | 1      |
+      | Group User Two | group_user_2@example.com | 1      |
+    And I am logged in as "Group User One"
+    And I am on "/group/add/closed_group"
```
Comment (jaapjan) [inline tests/behat/features/capabilities/group/group-create-closed.feature]: @frankgraave Could you adjust this to 

```
And I click "Add a group"
And I select "Closed group" (or something like this)
And I press "Continue"
```

File: tests/behat/features/capabilities/group/group-create-closed.feature [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    When I click "Test closed group"
+    Then I should see "About Group"
+    Then I should not see "Test closed group topic"
+    Then I should not see "Test closed group event"
+    And I should not see "This is a closed group post."
+    And I should not see "Stream"
+    And I should not see "Events"
+    And I should not see "Topics"
+    And I should not see "all upcoming events"
+    And I should not see "all topics"
```
Comment (jaapjan) [inline tests/behat/features/capabilities/group/group-create-closed.feature]: @frankgraave It is already a pretty thorough test. But could you also add some custom test steps for:
- Verifying that a non-member can not access the closed group topic
- Verifying that a non-member can not access the closed group event
- Verifying that a non-member can not access the closed group post

File: modules/social_features/social_group/config/install/views.view.group_topics.yml [hunk 4]
```diff
@@ -3,13 +3,13 @@ status: true
...
   config:
     - core.entity_view_mode.group_content.teaser
+    - group.content_type.closed_group-group_node-topic
     - group.content_type.open_group-group_node-topic
     - taxonomy.vocabulary.topic_types
   module:
     - group
     - node
     - taxonomy
-    - user
```
Comment (frankgraave) [inline modules/social_features/social_group/config/install/views.view.group_topics.yml]: @peterpolman Why is this removed?

File: modules/social_features/social_group/social_group.module [hunk 5]
```diff
@@ -75,11 +127,21 @@ function _social_group_group_open_group_edit_form_submit($form, FormStateInterfa
  *   - view_mode: View mode; e.g., 'full', 'teaser', etc.
  */
 function social_group_preprocess_group(&$variables) {
+  /** @var GroupInterface $group */
   $group = $variables['group'];
   $variables['joined'] = FALSE;
+  $variables['closed_group'] = FALSE;
+  $variables['cta'] = t('');
```
Comment (frankgraave) [inline modules/social_features/social_group/social_group.module]: Done.

File: modules/social_features/social_group/social_group.module [hunk 6]
```diff
@@ -460,6 +529,24 @@ function social_group_menu_local_tasks_alter(&$data, $route_name) {
   if (isset($data['tabs'][0]['group.edit_form'])) {
     unset($data['tabs'][0]['group.edit_form']);
   }
+
+  $user = \Drupal::currentUser();
+  // Get the Group object from the route
+  $group = Drupal::routeMatch()->getParameter('group');
+  // $group = _social_group_get_current_group(); ??
```
Comment (frankgraave) [inline modules/social_features/social_group/social_group.module]: Done.

File: modules/social_features/social_group/social_group.module [hunk 7]
```diff
@@ -48,6 +50,40 @@ function social_group_form_group_open_group_edit_form_alter(&$form, FormStateInt
...
+    $form['field_group_address']['widget'][0]['#title'] = '';
+  }
+}
+
+/**
+ * Implements hook_form_FORM_ID_alter().
+ */
+function social_group_form_group_closed_group_delete_form_alter(&$form, \Drupal\Core\Form\FormStateInterface $form_state, $form_id) {
+  // Add custom submit handler to delete all content of the group.
+//  drupal_set_message(t('are you sure you want to delete this group and all of the content?'), 'warning');
```
Comment (frankgraave) [inline modules/social_features/social_group/social_group.module]: Done.

File: modules/social_features/social_group/config/install/group.type.open_group.yml [hunk 8]
```diff
@@ -4,3 +4,5 @@ dependencies: {  }
 id: open_group
 label: 'Open group'
 description: 'This is a open group. Users may join without approval and all content added in this group will be visible for non members as well.'
+creator_roles:
```
Comment (frankgraave) [inline modules/social_features/social_group/config/install/group.type.open_group.yml]: Done.
